R/process_lookup_sc_demographics.R

```diff
@@ -29,29 +29,135 @@ process_lookup_sc_demographics <- function(
 
 
   # Data Cleaning ---------------------------------------
-
+  # TODO social care demographics - tidy up  code and make more efficient.
   sc_demog <- data %>%
     dplyr::mutate(
-      # use chi if upi is NA
-      upi = dplyr::coalesce(.data$upi, .data$chi_upi),
-      # check gender code - replace code 99 with 9
-      submitted_gender = replace(.data$submitted_gender, .data$submitted_gender == 99L, 9L)
-    ) %>%
-    dplyr::mutate(
+      # if one chi is missing then use the other
+      # TODO social care demographics - decide what to do with cases where last 4 digits of chi are different
+      chi_upi = ifelse(is.na(chi_upi), upi, chi_upi),
+      upi = ifelse(is.na(upi), chi_upi, upi),
+      submitted_date_of_birth = ifelse(is.na(submitted_date_of_birth), chi_date_of_birth, submitted_date_of_birth),
+      chi_date_of_birth = ifelse(is.na(chi_date_of_birth), submitted_date_of_birth, chi_date_of_birth),
+      chi_date_of_birth = lubridate::as_date(chi_date_of_birth),
+      submitted_date_of_birth = lubridate::as_date(submitted_date_of_birth),
+      # check gender code - replace code 99 with 9=
+
       # use CHI sex if available
+      # TODO social care demographics - check gender matches chi for extra validation check
+      submitted_gender = replace(.data$submitted_gender, .data$submitted_gender == 99L, 9L),
       gender = dplyr::if_else(
         is.na(.data$chi_gender_code) | .data$chi_gender_code == 9L,
         .data$submitted_gender,
         .data$chi_gender_code
-      ),
-      # Use CHI DoB if available
-      dob = dplyr::coalesce(.data$chi_date_of_birth, .data$submitted_date_of_birth)
+      )
     ) %>%
     # format postcodes using `phsmethods`
     dplyr::mutate(dplyr::across(
       tidyselect::contains("postcode"),
       ~ phsmethods::format_postcode(.x, format = "pc7")
-    ))
+    )) %>%
+    dplyr::distinct() %>%
+    # if only one option is available for chi then choose that
+    dplyr::mutate(chi = ifelse(chi_upi == upi | is.na(upi), chi_upi,
+      ifelse(is.na(chi_upi), upi, NA)
+    )) %>%
+    dplyr::mutate(
+      # if only one option is available for DOB then choose that
+      dob = ifelse(chi_date_of_birth == submitted_date_of_birth | is.na(submitted_date_of_birth), chi_date_of_birth,
+        ifelse(is.na(chi_date_of_birth), submitted_date_of_birth, NA)
+      ),
+      dob = lubridate::as_date(dob)
+    ) %>%
+    dplyr::arrange(chi, dob) %>%
+    dplyr::group_by(social_care_id, sending_location) %>%
+    tidyr::fill(chi, .direction = c("down")) %>%
+    tidyr::fill(dob, .direction = c("down")) %>%
```

Comment:
This work looks good, well done! I don’t fully understand the data and fill function yet, but is there a chance this could fill in a different chi or dob in a blank field if the row above belongs to a different person?

Reply:
So this is grouping by social_care_id and sending_location and only filling in chi and dob if one exists for that social_care_id and sending_location combination. 
This isn't great for combinations with two chis attached to them...but my thinking is that it will fix more cases than it will break...? This is another problem for another day.

Reply:
That makes sense, thank you!